Setup build benchmark jobs

breskeby opened the pull request:
Setting up automatic benchmark builds.

This PR allows testing impact on build changes on the build performance by comparing the PR source against master and run different scenarios to verify build performance by using the gradle profiler (see https://github.com/gradle/gradle-profiler) 

For now they can be triggered manually triggered by `run elasticsearch-ci/build-benchmark` from the PR

breskeby committed:
Add install profiler script for ci

breskeby committed:
configure new job for testing with personal git repo

breskeby committed:
configure new job for testing with personal git repo

breskeby committed:
Tweak build execution order

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

breskeby committed:
Fix default repository

breskeby committed:
Fix branch check for build performance build

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+pull-request+build-performance.yml:
```diff
@@ -0,0 +1,41 @@
+            JAVA8_HOME=$HOME/.java/java8
+            JAVA11_HOME=$HOME/.java/java11
+      - shell: |
+          #!/usr/local/bin/runbld --redirect-stderr
```
I think this first bit is a copy/paste error.

mark-vieira commented on .ci/scripts/install-gradle-profiler.sh:
```diff
@@ -0,0 +1,7 @@
+unzip $WORKSPACE/gradle-profiler-$PROFILER_VERSION.zip
+mv $WORKSPACE/gradle-profiler-$PROFILER_VERSION $WORKSPACE/gradle-profiler
+set -e
+$GRADLEW -S --max-workers=$MAX_WORKERS $@
```
I think we should efer to the `run-gradle.sh` script here given it's what does stuff like set MAX_WORKERS.

breskeby replied:
this changed. the gradle profiler doesn't use GRADLEW here as defined to run gradle.

mark-vieira commented on build-tools-internal/performance/elasticsearch-build-logic-change.scenarios:
```diff
@@ -0,0 +1,31 @@
+    warm-ups = 5
+    iterations = 10
+    git-checkout = {
+        build = "master"
```
I wonder if we should use the PR target branch here?

breskeby replied:
Don't bother with reviewing this yet. Still experimenting and playing around. Pushed to test ci job

breskeby replied:
I changed the scenarios to run 1) against the current commit and 2) against master for comparison reason

breskeby committed:
Fix casting in bootstrapping perf build tests

breskeby committed:
Update gradle profiler install script

breskeby committed:
Fix permissions on script

breskeby committed:
Cleanup install gradle profiler script

breskeby committed:
tweak performance build scenario

breskeby committed:
Fix passing gradle version to scenario file

breskeby committed:
Include gradle profile output in build tar

breskeby committed:
Update performance build

breskeby committed:
Update scenario title

breskeby committed:
Do not generate build tar for performance tests

breskeby committed:
rework build performance scenarios

breskeby committed:
Pass proper max workers to profiled gradle builds

breskeby committed:
Fix permissions for gradle profiler script

breskeby committed:
Fix profiler script

breskeby committed:
Add few scenarios to build performance test

breskeby committed:
More tweaks to scenario and build performance invocation

breskeby committed:
Remove abi change as it produces forbidden api failures

breskeby committed:
Wrap gradle profiler invocation

breskeby committed:
Tweak build properties

breskeby committed:
Rename performance to benchmark jobs

breskeby commented on .ci/scripts/run-gradle-profiler.sh:
```diff
@@ -0,0 +1,24 @@
+# drop page cache and kernel slab objects on linux
+[[ -x /usr/local/sbin/drop-caches ]] && sudo /usr/local/sbin/drop-caches
+rm -Rfv $WORKSPACE/gradle-user-home
+mkdir -p $WORKSPACE/gradle-user-home/init.d && cp -v $WORKSPACE/.ci/init.gradle $WORKSPACE/gradle-user-home
```
We configure the max worker here passed from the profiler to the gradle invocation

breskeby replied:
At the end comparing multiple benchmarks among each other doesn't make sense IMO. What we're interested here is just a comparison within the benchmark that is done in one run to see a tendency. I'd keep it to make as efficient as possible.

mark-vieira replied:
> At the end comparing multiple benchmarks among each other doesn't make sense IMO. What we're interested here is just a comparison within the benchmark that is done in one run to see a tendency. I'd keep it to make as efficient as possible.

Gotcha, and the "run" just compares HEAD against the PR changes so you get an idea if your change introduces a regression (or improvement), yes?

breskeby replied:
exactly

breskeby commented on .ci/jobs.t/elastic+elasticsearch+pull-request+build-benchmark.yml:
```diff
@@ -0,0 +1,52 @@
+            org-list:
+              - elastic
+            allow-whitelist-orgs-as-admins: true
+            trigger-phrase: '.*run\W+elasticsearch-ci/build-bench.*'
```
I think its enough to allow this to be triggered manually as we would not use it on a daily basis

breskeby replied:
fixed

breskeby commented on build-tools-internal/performance/elasticsearch-build-tool-update.scenarios:
```diff
@@ -0,0 +1,84 @@
+
+single_project_branch {
+    title = "single project (@testGitCommit@)"
+    cleanup-tasks = [":server:clean"]
```
Ideally we would test an ABI or non ABI change here. But the gradle profiler generated code for that, generates invalid formatted code (spotless and checkstyle complains). If we really want that we should fix that in the gradle profiler and provide a PR at gradle/gradle-profiler to make this work in our cases

mark-vieira replied:
Could we just run `spotlessApply` first?

breskeby replied:
Unfortunately we cannot invoke spotlessApply as part of the clean run as the abi change is done between the cleanup run and the benchmark run. If we include spotlessApply in the benchmark build it. 
For now I've added spotlessApply to the benchmark build. Since gradle is not yet running tasks in parallel within a project  `spotlessApply` should always be executed first without the need to tweak anything in our existing build logic. This tinkers a bit with the actual scenario as we usually never invoke `:server:spotlessApply :server:precommit` but that should probably be fine for now.

breskeby committed:
Cleanup build script and remove unused run benchmark script

breskeby commented on build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle:
```diff
@@ -9,8 +9,8 @@
 String buildNumber = System.getenv('BUILD_NUMBER')
-
-if (buildNumber) {
+String performanceTest = System.getenv('BUILD_PERFORMANCE_TEST')
```
we do not want to generate the tar.gz2 at the end of each build. we manually tar up the profile output instead and upload the results to google cloud

breskeby commented:
@mark-vieira @pugnascotia I think this is ready for a review.

pugnascotia commented on build-tools-internal/src/main/groovy/elasticsearch.build-complete.gradle:
```diff
@@ -9,8 +9,8 @@
-
-if (buildNumber) {
+String performanceTest = System.getenv('BUILD_PERFORMANCE_TEST')
+if (buildNumber && performanceTest != null) {
```
Do we know whether anyone is currently using these uploaded files e.g. to investigate ES issues? In which case, would they now need to run the build benchmarks to get these files?

mark-vieira replied:
Yeah, I think this condition is simply inverted. We want to generate the file when we are **not** running a performance test. So we should trigger when `performanceTest == null`.

breskeby replied:
fixed

breskeby committed:
Apply review feedback on pr job description

breskeby committed:
Change single project scenario

only run precommit task for now
run spotlessApply first to autofix autogenerated code

breskeby commented on .ci/jobs.t/elastic+elasticsearch+pull-request+build-benchmark.yml:
```diff
@@ -0,0 +1,52 @@
+            JAVA8_HOME=$HOME/.java/java8
+            JAVA11_HOME=$HOME/.java/java11
+      - shell: |
+         #!/usr/local/bin/runbld --redirect-stderr
```
fixed

mark-vieira commented on .ci/jobs.t/elastic+elasticsearch+pull-request+build-benchmark.yml:
```diff
@@ -0,0 +1,51 @@
+              - 6.8
+            excluded-regions:
+              - ^docs/.*
+            black-list-labels:
```
We can remove this. Doesn't make sense to have both a black list and a white list.

pugnascotia commented:
@breskeby did we ever document how this check is triggered?
